nautobot_golden_config/utilities/config_postprocessing.py

```diff
@@ -0,0 +1,177 @@
+"""Functions related to prepare configuration with postprocessing."""
+from functools import partial
+from typing import Optional
+import jinja2
+from jinja2 import exceptions as jinja_errors
+
+from django.http import HttpRequest
+from django.core.exceptions import ObjectDoesNotExist
+from django.utils.module_loading import import_string
+
+from nautobot.dcim.models import Device
+from nautobot.users.models import User
+from nautobot.extras.models.secrets import SecretsGroup
+from nautobot.extras.choices import SecretsGroupAccessTypeChoices
+
+from netutils.utils import jinja2_convenience_function
+
+from nautobot_golden_config import models
+from nautobot_golden_config.utilities.constant import PLUGIN_CFG, ENABLE_POSTPROCESSING
+from nautobot_golden_config.utilities.graphql import graph_ql_query
+from nautobot_golden_config.exceptions import RenderConfigToPushError
+from nautobot_golden_config.utilities.helper import get_device_to_settings_map
+
+
+def get_secret_by_secret_group_slug(
+    user: User,
+    secrets_group_slug: str,
+    secret_type: str,
+    secret_access_type: Optional[str] = SecretsGroupAccessTypeChoices.TYPE_GENERIC,
+) -> Optional[str]:
+    """Gets the secret from a Secret Group slug. To be used as a Jinja filter.
+
+    We assume that there is only one secret group corresponding to a model.
+
+    Args:
+        user (User): User object that performs API call to render push template with secrets.
+        secrets_group_slug (str): Secrets Group slug. It needs to be part of the GraphQL query.
+        secret_type (str): Type of secret, such as "username", "password", "token", "secret", or "key".
+        secret_access_type (Optional[str], optional): Type of secret such as "Generic", "gNMI", "HTTP(S)". Defaults to "Generic".
+
+    Returns:
+        Optional[str] : Secret value. None if there is no match. An error string if there is an error.
+    """
+    try:
+        secrets_group = SecretsGroup.objects.get(slug=secrets_group_slug)
+    except ObjectDoesNotExist:
+        return f"{secrets_group_slug} doesn't exist."
+
+    if not user.has_perm("extras.view_secretsgroup", secrets_group):
```

Comment:
This should be documented in the ADR.

Reply:
resolved in https://github.com/nautobot/nautobot-plugin-golden-config/commit/5b88c8afa091542b049f6de527ecb0d4001df50b